xarray/ufuncs.py

```diff
@@ -0,0 +1,125 @@
+"""xarray specific universal functions."""
+
+import textwrap
+import warnings
+
+import numpy as np
+
+import xarray as xr
+from xarray.core.groupby import GroupBy
+from xarray.namedarray.pycompat import array_type
+
+
+def _walk_array_namespaces(obj, namespaces):
+    if isinstance(obj, xr.DataTree):
+        for node in obj.subtree:
+            _walk_array_namespaces(node.dataset, namespaces)
+    elif isinstance(obj, xr.Dataset):
+        for name in obj.data_vars:
+            _walk_array_namespaces(obj[name], namespaces)
+    elif isinstance(obj, GroupBy):
+        _walk_array_namespaces(next(iter(obj))[1], namespaces)
+    elif isinstance(obj, xr.DataArray | xr.Variable):
+        _walk_array_namespaces(obj.data, namespaces)
+    elif isinstance(obj, array_type("dask")):
+        _walk_array_namespaces(obj._meta, namespaces)
+    else:
+        namespace = getattr(obj, "__array_namespace__", None)
+        if namespace is not None:
+            namespaces.add(namespace())
+
+    return namespaces
+
+
+class _UFuncDispatcher:
+    """Wrapper for dispatching ufuncs."""
+
+    def __init__(self, name):
+        self._name = name
+
+    def __call__(self, *args, **kwargs):
+        xps = set()
+        for arg in args:
+            _walk_array_namespaces(arg, xps)
+
+        xps.discard(np)
+        if len(xps) > 1:
+            names = [module.__name__ for module in xps]
+            raise ValueError(
+                f"Mixed array types {names} are not supported by xarray.ufuncs"
+            )
+
+        xp = next(iter(xps)) if len(xps) else np
+        func = getattr(xp, self._name, None)
+
+        if func is None:
+            warnings.warn(
+                f"Function {self._name} not found in {xp.__name__}, falling back to numpy",
+                stacklevel=2,
+            )
+            func = getattr(np, self._name)
+
+        return xr.apply_ufunc(func, *args, dask="parallelized", **kwargs)
```

Comment:
Is there ever a reason to use dask's ufuncs with `dask="allowed"` instead of the appropriate `_meta` array's namespace and `dask="parallelized"`? With `jax` for example, which doesn't have `__array_ufunc__`, this ends up converting to `numpy`. So it would have to be special cased.

Reply:
In user code using `xr.apply_ufunc` there is - `dask='allowed'` can be used to rechunk along a core dimension e.g. by applying a dask reduction ufunc along that dimension. Not sure if that's relevant here though.

Reply:
These are all elementwise so no core dimensions

Reply:
> Is there ever a reason to use dask's ufuncs with dask="allowed" instead of the appropriate _meta array's namespace and dask="parallelized"?

Yes, this feels like a cleaner solution to me.

> With jax for example, which doesn't have __array_ufunc__, this ends up converting to numpy. So it would have to be special cased

Is the concern here Dask wrapping JAX?

Generally, I think it's best for xarray to avoid introspecting into wrapped array types in Xarray, and leave nested wrapping up to other libraries, which can better understand their own implementation details. So Dask wrapping JAX should be fixed in Dask.

Reply:
> So Dask wrapping JAX should be fixed in Dask.

Totally fair. It looks like basically the same effort here would be required in dask then, because dask's [ufuncs](https://github.com/dask/dask/blob/main/dask/array/ufunc.py#L176) are all simple wrappers around the numpy version so they aren't aware of the namespace.